src/main/java/de/dennisguse/opentracks/services/GpsStatus.java

```diff
@@ -0,0 +1,217 @@
+        private boolean stopped = false;
+
+        @Override
+        public void run() {
+            if (gpsStatus != null && !stopped) {
+                if (!stopped) {
+                    onLocationChanged(null);
+                    gpsStatusHandler.postDelayed(gpsStatusRunner, getIntervalThreshold());
+                }
+            }
+        }
+
+        public void stop() {
+            stopped = true;
+        }
+    }
+
+    private Handler gpsStatusHandler = null;
+    private GpsStatusRunner gpsStatusRunner = null;
+
+    public GpsStatus(Context context, GpsStatusListener service) {
+        this.service = service;
+        this.context = context;
+        signalBadThreshold = PreferencesUtils.getRecordingDistanceInterval(context);
+        signalLostThreshold = getRecordingIntervalInMillis(PreferencesUtils.getMinRecordingInterval(context));
+        gpsStatusHandler = new Handler();
+    }
+
+    /**
+     * The service that uses GpsStatus has to call this method to stop the Runnable if needed.
+     */
+    public void stop() {
+        if (gpsStatusRunner != null) {
+            gpsStatusRunner.stop();
+            gpsStatusRunner = null;
+        }
+    }
+
+    /**
+     * Method to change the bad threshold from outside.
+     *
+     * @param value New preference value to signalBadThreshold.
+     */
+    public void onRecordingDistanceChanged(int value) {
+        signalBadThreshold = value;
+    }
+
+    /**
+     * Method to change the lost threshold from outside.
+     *
+     * @param value New preference value to signalLostThreshold.
+     */
+    public void onMinRecordingIntervalChanged(int value) {
+        signalLostThreshold = getRecordingIntervalInMillis(value);
+    }
+
+    /**
+     * @param value the preference value for recording interval.
+     * @return the preference value in milliseconds or SIGNAL_LOST_THRESHOLD_DEFAULT if the preference value is a default one.
+     */
+    private int getRecordingIntervalInMillis(int value) {
+        String minDefault = context.getString(R.string.min_recording_interval_default);
+        String minAdaptAccuracy = context.getString(R.string.min_recording_interval_adapt_accuracy);
+        String minAdaptBatteryLife = context.getString(R.string.min_recording_interval_adapt_battery_life);
+        if (value != Integer.valueOf(minDefault) &&
+                value != Integer.valueOf(minAdaptAccuracy) &&
+                value != Integer.valueOf(minAdaptBatteryLife)) {
+            return value * 1000; // in millis.
+        } else {
+            return SIGNAL_LOST_THRESHOLD_DEFAULT;
+        }
+    }
+
+    /**
+     * This method must be called from service every time a new location is received.
+     * Receive new location and calculate the new status if needed.
+     * Also, it'll run the runnable if signal is bad or stop it if the signal is lost.
+     */
+    public void onLocationChanged(final Location location) {
```

Comment:
Way too many ifs here.
Any chance to reduce this somehow?

Reply:
Ok, I simplify the GpsStatus removing one status because GPS_SIGNAL_OK and GPS_FIRST_FIX were the same.